src/io/s3_filesys.cc

```diff
@@ -192,6 +429,37 @@ std::string getEndpoint(std::string region_name) {
   }
 }
 
+/*!
+ * Encoding as required by SIG4
+ * \param str string to encode
+ * \param encodeSlash whether or not to encode slash (/) character
+ * \return
+ */
+std::string URIEncode(const std::string& str,
+                      bool encodeSlash = true) {
+  std::stringstream encoded_str;
+  encoded_str << std::hex << std::uppercase << std::setfill('0');
+  for (std::string::const_iterator it = str.begin(); it != str.end(); ++it) {
+    char c = *it;
+    if ((c >= 'a' && c <= 'z') ||
+        (c >= 'A' && c <= 'Z') ||
+        (c >= '0' && c <= '9') ||
+        c == '-' || c == '_' ||
+        c == '.' || c == '~') {
+      encoded_str << c;
+    } else if (c == '/') {
+      if (encodeSlash) {
+        encoded_str << "%2F";
+      } else {
+        encoded_str << c;
+      }
+    } else {
+      encoded_str << '%';
+      encoded_str << std::setw(2) << static_cast<unsigned>(c);
```

Comment:
Wont this just output % followed by decimal value ? Dont you want the hexadecimal representation ?

Reply:
No, because of the usage of std::hex. Which it is set as basefield, integer values inserted into the stream are expressed in hexadecimal base.

An example
```
#include <iostream>  
int main () {
  int n = 70;
  std::cout << std::dec << n << '\n';
  std::cout << std::hex << n << '\n';
  std::cout << std::oct << n << '\n';
  return 0;
}
```
Output
```
70
46
106
```

Reply:
Thanks for explaining. I missed that line.